go-controller/pkg/ovn/port_cache.go

```diff
@@ -0,0 +1,84 @@
+package ovn
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"time"
+)
+
+type portCache struct {
+	sync.RWMutex
+
+	stopChan <-chan struct{}
+	cache    map[string]*lpInfo
+}
+
+type lpInfo struct {
+	name          string
+	uuid          string
+	logicalSwitch string
+	ip            net.IP
+	mac           net.HardwareAddr
+	// dead indicates that this object should not be requeued for deletion
+	dead bool
+}
+
+func newPortCache(stopChan <-chan struct{}) *portCache {
+	return &portCache{
+		stopChan: stopChan,
+		cache:    make(map[string]*lpInfo),
+	}
+}
+
+func (c *portCache) get(logicalPort string) (*lpInfo, error) {
+	c.RLock()
+	defer c.RUnlock()
+	if info, ok := c.cache[logicalPort]; ok {
+		return info, nil
+	}
+	return nil, fmt.Errorf("logical port %s not found in cache", logicalPort)
+}
+
+func (c *portCache) add(logicalSwitch, logicalPort, uuid string, mac net.HardwareAddr, ip net.IP) *lpInfo {
+	c.Lock()
+	defer c.Unlock()
+	portInfo := &lpInfo{
+		logicalSwitch: logicalSwitch,
+		name:          logicalPort,
+		uuid:          uuid,
+		ip:            ip,
+		mac:           mac,
+	}
+	c.cache[logicalPort] = portInfo
+	return portInfo
+}
+
+func (c *portCache) remove(logicalPort string) {
+	c.Lock()
+	defer c.Unlock()
+	info, ok := c.cache[logicalPort]
+	if !ok || info.dead {
+		return
+	}
+	info.dead = true
+
+	// Removal must be deferred since, for example, NetworkPolicy pod handlers
+	// may run after the main pod handler and look for items in the cache
+	// on delete events.
+	go func() {
+		select {
+		case <-time.After(time.Minute):
+			c.Lock()
+			defer c.Unlock()
+			// recheck whether the port info is dead since it could
+			// have been resurrected if the pod was re-added in the
+			// last 10 seconds
+			if info, ok := c.cache[logicalPort]; ok && info.dead {
```

Comment:
It is now `in the last one minute` instead of `last 10 seconds`.

I think this will work as long as we serialize all the events related to a particular pod and that each handler does its job inside that callback. If the handlers do call a go routine, then they must get a reference to the portInfo as the argument.

Reply:
@girishmg fixed the comment and added some code to make sure we only remove the port info if it wasn't re-added during the timer.